nativelib/src/main/resources/gc/commix/Heap.c

```diff
@@ -0,0 +1,542 @@
+             object++) {
+            assert(!ObjectMeta_IsMarked(object));
+        }
+
+        current = next;
+        lineMetas = nextLineMetas;
+        currentBlockStart = nextBlockStart;
+    }
+    assert(current == limit);
+}
+#endif
+
+void Heap_Collect(Heap *heap) {
+#ifdef ENABLE_GC_STATS
+    Stats *stats = heap->stats;
+    if (stats != NULL) {
+        stats->collection_start_ns = scalanative_nano_time();
+    }
+#else
+    Stats *stats = NULL;
+#endif
+    assert(Sweeper_IsSweepDone(heap));
+#ifdef DEBUG_ASSERT
+    Heap_clearIsSwept(heap);
+    Heap_assertIsConsistent(heap);
+#endif
+    heap->mark.lastEnd_ns = heap->mark.currentEnd_ns;
+    heap->mark.currentStart_ns = scalanative_nano_time();
+    Marker_MarkRoots(heap, stats);
+    heap->gcThreads.phase = gc_mark;
+    // make sure the gc phase is propagated
+    atomic_thread_fence(memory_order_release);
+    // start main gc thread (0)
+    sem_post(&heap->gcThreads.start0);
+    while (!Marker_IsMarkDone(heap)) {
+        Marker_Mark(heap, stats);
+        if (!Marker_IsMarkDone(heap)) {
+            sched_yield();
+        }
+    }
+    heap->gcThreads.phase = gc_idle;
+    heap->mark.currentEnd_ns = scalanative_nano_time();
+#ifdef ENABLE_GC_STATS
+    if (stats != NULL) {
+        Stats_RecordEvent(stats, event_mark, heap->mark.currentStart_ns,
+                          heap->mark.currentEnd_ns);
+    }
+#endif
+    Heap_Recycle(heap);
+}
+
+bool Heap_shouldGrow(Heap *heap) {
+    uint32_t freeBlockCount = (uint32_t)blockAllocator.freeBlockCount;
+    uint32_t blockCount = heap->blockCount;
+    uint32_t recycledBlockCount = (uint32_t)allocator.recycledBlockCount;
+    uint32_t unavailableBlockCount =
+        blockCount - (freeBlockCount + recycledBlockCount);
+
+#ifdef DEBUG_PRINT
+    printf("\n\nBlock count: %" PRIu32 "\n", blockCount);
+    printf("Unavailable: %" PRIu32 "\n", unavailableBlockCount);
+    printf("Free: %" PRIu32 "\n", freeBlockCount);
+    printf("Recycled: %" PRIu32 "\n", recycledBlockCount);
+    fflush(stdout);
+#endif
+
+    uint64_t timeInMark = heap->mark.currentEnd_ns - heap->mark.currentStart_ns;
+    uint64_t timeTotal = heap->mark.currentEnd_ns - heap->mark.lastEnd_ns;
+
+    return timeInMark >= GROWTH_MARK_FRACTION * timeTotal || freeBlockCount * 2 < blockCount ||
+           4 * unavailableBlockCount > blockCount;
+}
+
+void Heap_Recycle(Heap *heap) {
+    Allocator_Clear(&allocator);
+    LargeAllocator_Clear(&largeAllocator);
+    BlockAllocator_Clear(&blockAllocator);
+
+    // use the reserved block so mutator can does not have to lazy sweep
+    // but can allocate imminently
+    BlockAllocator_UseReserve(&blockAllocator);
+
+    // all the marking changes should be visible to all threads by now
+    atomic_thread_fence(memory_order_seq_cst);
+
+    // before changing the cursor and limit values, makes sure no gc threads are
+    // running
+    GCThread_JoinAll(heap);
```

Comment:
This looks like you're going to wait here *on the mutator thread*, is there any way to avoid that?